Added pegboard wip

Derek-Strasters opened the pull request:
Let me know if this is how you imagined the _wip file to work, or how you'd like it changed 👍

stohrendorf reviewed (changes requested):
Please squash all changes into a single commit.

Derek-Strasters committed:
adding metro_control_panel_001_pegboard_wip.py
removing non-essential .idea files

allowing watcher tasks for formatting

re-adding watcherTasks.xml

separating experimentation into a _wip file

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -32,3 +34,29 @@ def _swap_pegs(pegs: Pegs, a: str, b: str) -> Pegs:
+
+
+def _markers_to_bool_list(peg_str: str) -> Iterable[bool]:
+    return map(lambda b: True if b == "M" else False, peg_str)
```
```suggestion
    return peg == "M" for peg in peg_str
```

Derek-Strasters replied:
Good call, more pythonic that way

Derek-Strasters replied:
I am changing them to return lists so that they are reusable, but I am removing "list" from the name anyways since I don't think it adds to readability.

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -32,3 +34,29 @@ def _swap_pegs(pegs: Pegs, a: str, b: str) -> Pegs:
+
+
+def _pegs_to_bool_list(peg_str: str) -> Iterable[bool]:
+    return map(lambda b: False if b == "E" else True, peg_str)
```
```suggestion
    return peg != "E" for peg in peg_str
```

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -32,3 +34,29 @@ def _swap_pegs(pegs: Pegs, a: str, b: str) -> Pegs:
+    def nor(a: bool, b: bool) -> bool:
+        return not (a or b)
+
+    def pass_thru(a: bool, b: bool) -> bool:
+        return a
+
+    return [nor if bit else pass_thru for bit in _markers_to_bool_list(peg_str)]
```
```suggestion
    def left(a: bool, b: bool) -> bool:
        return a

    return [nor if bit else left for bit in _markers_to_bool_list(peg_str)]
```

Derek-Strasters replied:
Great suggestion, names are always tricky

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -32,3 +34,29 @@ def _swap_pegs(pegs: Pegs, a: str, b: str) -> Pegs:
     )
+
+
+def _markers_to_bool_list(peg_str: str) -> Iterable[bool]:
```
That name is wrong, it's not return a `list`.

Derek-Strasters replied:
Good catch.  It probably doesn't make sense to have types in the name anyways since we are using type hints

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -32,3 +34,29 @@ def _swap_pegs(pegs: Pegs, a: str, b: str) -> Pegs:
+    return map(lambda b: True if b == "M" else False, peg_str)
+
+
+def _pegs_to_bool_list(peg_str: str) -> Iterable[bool]:
```
This name is also wrong.

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -32,3 +34,29 @@ def _swap_pegs(pegs: Pegs, a: str, b: str) -> Pegs:
+    return map(lambda b: False if b == "E" else True, peg_str)
+
+
+def _invert_bool_list(bits: Iterable[bool]) -> Iterable[bool]:
```
Also this name.

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -32,3 +34,29 @@ def _swap_pegs(pegs: Pegs, a: str, b: str) -> Pegs:
+
+
+def _invert_bool_list(bits: Iterable[bool]) -> Iterable[bool]:
+    return map(lambda b: not b, bits)
```
```suggestion
    return not b for b in bits
```

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -32,3 +34,29 @@ def _swap_pegs(pegs: Pegs, a: str, b: str) -> Pegs:
+
+
+def _bools_to_binary_str(bits: Iterable[bool]) -> str:
+    return insert_spaces("".join(str(int(b)) for b in bits), 8)
```
```suggestion
    return insert_spaces("".join("1" if b else "0") for b in bits), 8)
```

Also, maybe make this generally available, add some docs and doctests.

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,136 @@
+        b_ = _pegs_to_bool_list(b)
+        transformed = [op(first, second) for first, op, second in zip(a_, mask_, b_)]
+        bytes_str = binary_decode(_bools_to_binary_str(transformed))
+        return " ".join(char if 126 > ord(char) > 33 else str(ord(char)) for char in bytes_str)
```
I don't like the output format. Maybe change it to something like `0x{code}={char}` if it's a non-control ascii code, otherwise just `0x{code}`.

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,136 @@
+        mask_ = _bools_to_nor_mask(mask)
+        b_ = _pegs_to_bool_list(b)
+        transformed = [op(first, second) for first, op, second in zip(a_, mask_, b_)]
+        bytes_str = binary_decode(_bools_to_binary_str(transformed))
```
Converting a sequence of booleans to a string and then back to a sequence of bytes seems a bit complicated.

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,136 @@
+    print()
+    for panel in metro_control_panel_001_pegboard:
+        # bottom nor top from top markers inverted
+        print(transform_pegs_all_inverted(panel[1],
+                                          panel[0],
+                                          panel[0]))
```
I'm a bit confused these are on separate lines. Did you run the autoformatters? What do the linters say?

Derek-Strasters replied:
The reason I fight the linter on this one is that I believe this has better readability.  Especially since the same function is being called many times, having the arguments physically grouped makes it easier to spot check what line is getting what inputs to help avoid errors.

This is subjective though and I would happily change it.

That said I have some changes coming in that make everything more pretty, satisfying the linter and readability.

stohrendorf replied:
The reason for the formatter and linter is to avoid any discussions about these things. The automatic formatting is the truth, never fight it. If you want to split it into separate lines, add a trailing comma, this forces black to split it.

Derek-Strasters replied:
I hadn't used Black before, good to know this trick, thanks!

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,136 @@
+
+
+if __name__ == "__main__":
+    def transform_pegs(a: str, mask: str, b: str) -> str:
```
Adding some docs to each of the transformations what they actually do would be nice. All these functions look similar, too, maybe refactor it to extract common parts.

Derek-Strasters replied:
Will do!

Derek-Strasters commented:
Thanks for a really good code review!  I'll get these changes in.

Derek-Strasters committed:
fix naming conventions
fix return value schemes
remove unneeded maps
general refactor and cleanup

Derek-Strasters committed:
fix naming conventions
fix return value schemes
remove unneeded maps
general refactor and cleanup

Update textures/metro_control_panel_001_pegboard_wip.py

Better naming convention

Co-authored-by: Steffen Ohrendorf <steffen.ohrendorf@gmx.de>

implemented termcolor

reduced redundant functions
added docstrings
fixed doctests
fixed peg group name hack
better naming convention inverses -> inverted

Better naming convention: chunks -> chunked

Adding termcolor

pegboard_experiments

Replaced named tuple with the faster dataclass
optimized imports

changed variable name for clarity

Removed cruft and unused/redundant function

replaced function with typed lambda

replaced function signature with something more generic

Derek-Strasters commented:
This should be  a lot cleaner now, some of the results are interesting too.  There is one combination that has all three bytes as alphanumeric.  I think I'll explore UTF-8 encoding next.

Also, I've got another thing that I've been working on, I'll check it in to a separate branch so that you can look at.

edit: https://github.com/Derek-Strasters/infra-scratchpad/tree/steganography/images

Derek-Strasters commented:
Whoops, I am just realizing I accidentally duplicated one of your methods in utils.py, one moment and I'll have that fixed.

stohrendorf commented on infra/utils.py:
```diff
@@ -1,12 +1,32 @@
 Indexable = Union[List[T], Union[T], str]
 
 
+def chunks(items: Collection[T], n: int, chunk_type: Type[Collection[T]] = None) -> Iterable[Collection[T]]:
```
Please remove the `chunk_type` parameter - it makes things complicated and people can do stuff like `[set(x) for x in chunked(...)] anyways`. It also adds stuff to the functionality that's beyond splitting a sequence into chunks.

Also, to feel the same as `sorted`, please call it `chunked`.

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -1,8 +1,15 @@
 Pegs = Tuple[Tuple[str, str], ...]
+PegGroup = NamedTuple("PegGroup", [("name", str), ("pegs", Sequence[bool]), ("markers", Sequence[bool])])
+
+# █◉█oo█◉█◉█ooo█◉█◉█◉ooo█◉ | █◉█◉█◉oooo█◉█◉oooo█◉█◉█◉ | █◉█◉█◉█◉oooooooooooooooo
```
please avoid using unicode characters - I know it's tempting, but it makes things overly complicated. One has to use a unicode lookup thingie to find the character and then paste it here. Doing the reverse, i.e. finding the unicode code point for a unicode character, is even harder. If you want to use unicode, please add constants and unicode code point escape sequences.

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -25,10 +32,95 @@ def _remove_empty(pegs: Pegs) -> Pegs:
+    return tuple(peg == "M" for peg in peg_str)
+
+
+def pegs_to_bools(peg_str: str) -> Tuple[bool]:
```
these two functions basically do the same. Maybe use something like that:

```
def sequence_elements_equal(sequence: Sequence[T], needle: T) -> Tuple[bool]:
    return tuple(element == needle for element in sequence)
```

(just a blind guess, haven't tested it)

Derek-Strasters replied:
Good call!

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -25,10 +32,95 @@ def _remove_empty(pegs: Pegs) -> Pegs:
+    def nor(a: bool, b: bool) -> bool:
+        return not (a or b)
+
+    # noinspection PyUnusedLocal
```
it _is_ used, so why the suppression?

Derek-Strasters replied:
![image](https://user-images.githubusercontent.com/10388331/138060456-867d2b53-a076-41d9-86f4-bac068e33386.png)
My PyCharm seems to have an issue with the bool "b" not being used in the function, even though this is the rare case where it is required to have that signature for later use.  I could work around this by using lambdas if you'd rather not have a suppression statement in the code base.

stohrendorf replied:
That's very strange. I can live with it, but I'd prefer it wouldn't be necessary without convoluting the code.

Derek-Strasters replied:
I found the solution!  And I didn't have to resort to something as un-pythonic as a named, typed, lambda:

```    
def left(*args) -> bool:
        return args[0]
```

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -25,10 +32,95 @@ def _remove_empty(pegs: Pegs) -> Pegs:
+        return a
+
+    def nor_mask_function(left_: Collection[bool], right_: Collection[bool]) -> Tuple[bool]:
+        assert len(left_) == len(right_) == len(masking_bools)
```
that's an interesting syntax, didn't know about it (if it works)

Derek-Strasters replied:
I love Python!  Endless quirks and niceties.  You couldn't pay me enough to go back to Java hahaha!

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -25,10 +32,95 @@ def _remove_empty(pegs: Pegs) -> Pegs:
+    def nor_mask_function(left_: Collection[bool], right_: Collection[bool]) -> Tuple[bool]:
+        assert len(left_) == len(right_) == len(masking_bools)
+
+        operators = [nor if bit else left for bit in masking_bools]
```
no need to convert it to a list here as it is used only in a `zip` statement

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -25,10 +32,95 @@ def _remove_empty(pegs: Pegs) -> Pegs:
+    def left(a: bool, b: bool) -> bool:
+        return a
+
+    def nor_mask_function(left_: Collection[bool], right_: Collection[bool]) -> Tuple[bool]:
```
```suggestion
    def nor_mask_function(a: Collection[bool], b: Collection[bool]) -> Tuple[bool]:
```

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -25,10 +32,95 @@ def _remove_empty(pegs: Pegs) -> Pegs:
+    return nor_mask_function
+
+
+def bools_to_int(bits: Collection) -> int:
```
```suggestion
def bools_to_int_be(bits: Collection) -> int:
```

Derek-Strasters replied:
Ah, for big-endian 👍

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,85 @@
+
+
+def is_alpha_num(char: str) -> bool:
+    return any(("0" <= char <= "9", "A" <= char <= "Z", "a" <= char <= "z"))
```
I'm not sure, but I _think_ I have seen something like `isalnum` somewhere in the standard library. If it exists, please use it here. Also, this function seems wrong, as `is_alpha_num("-.ä-,-.,-.,9")` would evaluate to `True` as you're using the `any` aggregation.

Derek-Strasters replied:
Heh, apparently is a direct member of the str and bytes class.  Not sure how I never noticed that before!

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,85 @@
+
+    def color_red(chars: str) -> str:
+        """Colorize alphanumeric characters"""
+        return "".join(f"\x1b[11;31m{char}\x1b[0m" if is_alpha_num(char) else char for char in chars)
```
if you want console colors, please add colorama or something like that and don't do custom escape sequences ;)

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,85 @@
+        return "".join(f"\x1b[11;31m{char}\x1b[0m" if is_alpha_num(char) else char for char in chars)
+
+    def bools_to_ascii_str(bits: Sequence[bool]) -> str:
+        # assert len(bits) % 8 == 0
```
dead code, please remove it

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,85 @@
+    # Flatten the peg groupings for easier handling
+    # pegs[1,1]                 pegs[3,1]    pegs[4,1]
+    # pegs[1,2]                 pegs[3,2]    pegs[4,2]
+    peg_groups = []
+    for i, panels in enumerate(metro_control_panel_001_pegboard):
+        for j, peg_str in enumerate(panels):
+            peg_group = PegGroup(
+                # Correcting for the blank panel
+                name=f"pegs[{i + (1 if not i else 2)},{j + 1}]",
+                pegs=pegs_to_bools(peg_str),
+                markers=markers_to_bools(peg_str),
+            )
+            peg_groups.append(peg_group)
```
This can be converted into a list comprehension.

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -1,8 +1,15 @@
 
+NorMask = Callable[[Sequence[bool], Sequence[bool]], Tuple[bool]]
 Pegs = Tuple[Tuple[str, str], ...]
+PegGroup = NamedTuple("PegGroup", [("name", str), ("pegs", Sequence[bool]), ("markers", Sequence[bool])])
```
No... don't use `NamedTuple`, please, and use `@dataclass` instead.

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,85 @@
+        for j, peg_str in enumerate(panels):
+            peg_group = PegGroup(
+                # Correcting for the blank panel
+                name=f"pegs[{i + (1 if not i else 2)},{j + 1}]",
```
If the panel data is wrong, make it correct so it works, and don't "fix" it by writing `1 if not i else 2`.

Derek-Strasters replied:
Fixed!

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,85 @@
+    print()
+    print(" " * 27 + (" " * 8).join(f"col {x: <2d}" for x in range(1, 9)))
+
+    # Iterate through all combinations of the peg groupings
+    # Bits are represented by peg locations
+    # A nor mask is made from the marker locations from one of the peg groupings
```
Please separate multiple sentences by a dot.

Derek-Strasters replied:
Done. 👍

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,85 @@
+    # A nor mask is made from the marker locations from one of the peg groupings
+    for i, pegs1 in enumerate(peg_groups):
+        mask_1 = bools_to_nor_mask(pegs1.markers)
+        inverses_1 = invert_bools(pegs1.pegs)
```
```suggestion
        inverted_1 = invert_bools(pegs1.pegs)
```

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,85 @@
+        for j, pegs2 in enumerate(peg_groups):
+
+            mask_2 = bools_to_nor_mask(pegs2.markers)
+            inverses_2 = invert_bools(pegs2.pegs)
```
```suggestion
            inverted_2 = invert_bools(pegs2.pegs)
```

stohrendorf commented on textures/metro_control_panel_001_pegboard_wip.py:
```diff
@@ -0,0 +1,85 @@
+
+            print(f"{pegs1.name} x {pegs2.name}", end=" =  ")
+
+            print(bools_to_ascii_str(mask_1(pegs1.pegs, pegs2.pegs)), end="   ")  # col 1
+            print(bools_to_ascii_str(mask_2(pegs1.pegs, pegs2.pegs)), end="   ")  # col 2
+
+            print(bools_to_ascii_str(mask_1(pegs1.pegs, inverses_2)), end="   ")  # col 3
+            print(bools_to_ascii_str(mask_2(pegs1.pegs, inverses_2)), end="   ")  # col 4
+
+            print(bools_to_ascii_str(mask_1(inverses_1, pegs2.pegs)), end="   ")  # col 5
+            print(bools_to_ascii_str(mask_2(inverses_1, pegs2.pegs)), end="   ")  # col 6
+
+            print(bools_to_ascii_str(mask_1(inverses_1, inverses_2)), end="   ")  # col 7
+            print(bools_to_ascii_str(mask_2(inverses_1, inverses_2)), end="   ")  # col 8
```
If you want to display something tabular, investigate whether `tabulate` is appropriate, or if something else fits even better to display tabular data.

Derek-Strasters replied:
I'll look into it, thanks for the tip!

Derek-Strasters replied:
After having thought about it, I originally laid this out here as a place to quickly test things in a way that I could get a glancing look to see if anything intelligible was coming out.

Ultimately I'd like to integrate your natural language processing tools in with a bunch of random transforms that can be tried in a similar nondeterministic form to some of the other things you've produced.

In the mean time, I do still think it would be fun to play with some tabulation as I haven't had the pleasure to work with that lib before, so I'll try to get another commit in with that soon.

stohrendorf reviewed (changes requested):
Sorry it did take so long, I had a PC hardware fault :'(

Derek-Strasters commented:
Sorry to hear about your computer problems!  Glad to hear you got it worked out.  I should have some changes coming in soon, as well I've got a few other big things I've been working on, probably big enough that I'll push them into separate branches for later review so that this PR doesn't get too messy 👍

stohrendorf commented:
Keep in mind that everything I say is up for discussion.

Derek-Strasters committed:
naming convention update

unnecessary list removed

stohrendorf commented on pyproject.toml:
```diff
@@ -6,6 +6,7 @@ authors = ["Your Name <you@example.com>"]
 
 [tool.poetry.dependencies]
 python = "^3.8"
+termcolor = "^1.1.0"
```
```suggestion
termcolor = "*"
```

stohrendorf commented on infra/utils.py:
```diff
@@ -1,12 +1,29 @@
 Indexable = Union[List[T], Union[T], str]
 
 
+def chunked(items: Sequence[T], n: int) -> Sequence[Sequence[T]]:
```
some check that `n >= 1` would be nice

stohrendorf commented on infra/utils.py:
```diff
@@ -1,12 +1,29 @@
-
-from typing import Iterable, List, Sequence, Tuple, TypeVar, Union
+import types
+from typing import Iterable, List, Sequence, Tuple, TypeVar, Union, Type, Any, Callable, cast, Collection
```
hm. isort should have taken care of sorting it alphabetically.

stohrendorf commented on textures/metro_control_panel_001_pegboard.py:
```diff
@@ -25,10 +40,83 @@ def _remove_empty(pegs: Pegs) -> Pegs:
+    """
+    Creates an elementwise function that can called with two lists of booleans
+    of the same shape.
+    How the returned nor mask function is meant to be used:
+    (A, B, and C are lists; * is the nor mask function)
+    A * B = C
+    For C[n] the value A[n] is given where masking_bools[n] is False
+    and A[n] NOR B[n] is given where masking_bools[n] is True.
```
I think something like that would be more clear (although it doesn't match the implementation):

```suggestion
    The function returns a sequence of binary functions. Depending on
    masking_bools[i], it returns either nor(a,b) if True, or left(a,b):=a otherwise.
    E.g., bools_to_nor_mask((True, False)) returns (nor, left) ...
```

Derek-Strasters replied:
This was a great suggestion, I sometimes struggle to get the wording right.

Derek-Strasters commented:
Hey, just checking in. Sorry I haven't been very active.  I've been dealing with some unfortunate medical issues.  Looking forward to jumping back in.  I've been developing a  "Bits" class that I think could come in handy in multiple places as well as cleaning up what's been done so far.

It's on this branch if you'd like to have a look: 
https://github.com/Derek-Strasters/infra-scratchpad/blob/bits_utility/infra/utils.py

stohrendorf commented:
Yes, I've seen it, although it seems rather complicated and overengineered at a first glimpse, I'm curious where this is going.

Derek-Strasters commented:
I definitely feel like there is a way (that I'm not seeing) to reach my goals that is simpler and more elegant.  But I really wanted a class that I can iterate over for Booleans, can add and remove Boolean values, and that can hold large amounts of data efficiently - specifically for use with some image and audio processing stuff I've been working on.

I will say I am probably a bit guilty of premature optimization by not first backing the class with a simple list of Booleans.  But in any case it's nearly ready for use after some clean up.

Derek-Strasters committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	infra/utils.py
#	textures/metro_control_panel_001_pegboard.py
#	textures/metro_control_panel_001_pegboard_wip.py

Derek-Strasters committed:
refactoring and updating documentation

Derek-Strasters commented:
I'm noticing flake8 seems to consistently be making a stink about import order, I tried adding `application-import-names = "infra other textures audio"` to the .flake8 config but it still gets some of the application imports wrong half the time.  I was wondering if perhaps you'd be interested in removing flake8-import-order from the dev-dependencies and let the linting stack defer to what isort prescribes?  Alternatively I noticed there is a flake8-isort plugin (and it even supports pyproject.toml).

stohrendorf commented:
I though I had added flake8-isort, that's obviously a mistake I made, which now is complaining
